### ISSUE TITLE
support multiple per-row updates in a single mutation

### ISSUE THREAD

**Issue by 0x777** [hasura/graphql-engine#2768]
Currently we allow updating multiple rows (through `where`) but all of them will get the same update. We need to add support for cases where the updates are different for each row, say you want to set `name` to `Hello` for the row with `id=1` and to `World` when `id=2`. Something like this maybe? 
```graphql
{
  update_user_many(
    updates: [
      { where: {id: {_eq: 1}, _set: {name: "hello"}},
      { where: {id: {_eq: 2}, _set: {name: "world"}}
    ]
  ) {
    affected_rows
  }
}
```
Notes:
1. What happens when there is overlap across the `where` conditions? What would `affected_rows` and `returning` return?
2. Constructing the `updates` argument would need a bit of a boilerplate.

Maybe we can simplify the api to just use the primary key/unique constraints?
```graphql
{
  update_user_many(
    updates: [
      { id: 1, _set: {name: "hello"}},
      { id: 2, _set: {name: "world"}}
    ]
  ) {
    affected_rows
  }
}
```
We can probably use `update .. from` as suggested here: https://stackoverflow.com/a/18799497

**Comment by revskill10** [comment]
@0x777 As i see, for the simpler API , we could also use `where` with unique constraint key ?

**Comment by hutber** [comment]
I would enjoy this feature

**Comment by levid** [comment]
I would also really like this feature. Currently I have to update multiple records individually and this would be so much easier and more efficient.

**Comment by marcfalk** [comment]
Would love this too! For now I'm using [upsert](https://docs.hasura.io/1.0/graphql/manual/mutations/upsert.html#update-selected-columns-on-conflict) although it's not highly recommended in the docs. What are the drawbacks of doing that until a multi-update feature is here?

**Comment by revskill10** [comment]
Currently, i use multiple mutation in one graphql query to achieve this, as Hasura allows you to use multiple mutation inside one query, and all of them will execute in one transaction.

**Comment by FarazPatankar** [comment]
Is there an update on this at all? Also, I am considering going the upsert way like @marcfalk has and was wondering if there are any drawbacks as well. Would love some thoughts on this. @tirumaraiselvan

**Comment by xiaoyu-tamu** [comment]
> Is there an update on this at all? Also, I am considering going the upsert way like @marcfalk has and was wondering if there are any drawbacks as well. Would love some thoughts on this. @tirumaraiselvan

The pk may leaving gaps. 

e.g.
id 1
id 9
id 100

**Comment by FarazPatankar** [comment]
> > Is there an update on this at all? Also, I am considering going the upsert way like @marcfalk has and was wondering if there are any drawbacks as well. Would love some thoughts on this. @tirumaraiselvan
> 
> The pk may leaving gaps unless you use uuid
> 
> e.g.
> id 1
> id 9
> id 100

Anything other than this? And does this have any bad effects apart from the fact that there are gaps?

**Comment by aaronbski** [comment]
+1

**Comment by FarazPatankar** [comment]
For anyone struggling with this, I ended up using the upsert mutation for this due to the lack of a response and it works perfectly.

**Comment by xiaoyu-tamu** [comment]
Is this going to implement? I don't see why cron/schedule job has higher priority than multiple per-row updates and multiple auth role.

**Comment by hafiztahajamil** [comment]
@praveenweb This is an important feature. You must assign this to someone. Thanks

**Comment by BPiepmatz** [comment]
would love to see this too 👍

**Comment by hafiztahajamil** [comment]
@praveenweb Any updates ?

**Comment by hafiztahajamil** [comment]
> Currently, i use multiple mutation in one graphql query to achieve this, as Hasura allows you to use multiple mutations inside one query, and all of them will execute in one transaction.

@revskill10 How to dynamically add multiple mutations to a single one each with different variable values? How to make use of the aliases dynamically

**Comment by hafiztahajamil** [comment]
Can you please give an example on how to do it  ?

**Comment by revskill10** [comment]
@hafiztahajamil For example, i want to update `first_name` of multiple users in one mutation, i would do:

```
mutation {
update_users_1(objects:users1_objects) { affected_rows }
update_users_2(objects:users2_objects) { affected_rows }
update_users_3(objects:users3_objects) { affected_rows }
update_users_4(objects:users4_objects) { affected_rows }
}
```
In above mutation, i generated `update_users_X` from code , as well as `usersX_objects`.

**Comment by valstu** [comment]
Any updates to this, currently most of my updates need to fallback to upserts and that is something I wouldn't like to use.

**Comment by Lexe003** [comment]
Updating multiple rows with different Pk's by calling the API multiple times sounds really bad. Hope this will be implemented soon. This should be a core feature.

**Comment by harpyng** [comment]
Please provide an update on this feature!

Although the upsert alternative is a possibility the documentation specifically suggests otherwise:
https://hasura.io/docs/latest/graphql/core/databases/postgres/mutations/upsert.html
"Upsert is not a substitute for update¶
The upsert functionality is sometimes confused with the update functionality. However, they work slightly differently. An upsert mutation is used in the case when it’s not clear if the respective row is already present in the database. If it’s known that the row is present in the database, update is the functionality to use.

For an upsert, all columns that are necessary for an insert are required."

I have a use case where I want to update JSON fields across multiple ecommerce products at once and doing it by running a mutation with hundreds of individual updates or serially calling the API is not ideal.

Upsert is a possibility, but as stated above from the docs, it requires all columns necessary for an insert - so doesn't work well if there are not null constraints to consider. (or requires a lot of additional and unnecessary information to be sent with each mutation).

**Comment by noel-dolan** [comment]
Wow!  First posted in 23 Aug 2019 and still not a feature!  This is exactly what I was/am looking for too, I won't hold my breath! :p  upsert really isn't an option as all fields are required which would be other data being wiped.  Surely this should be high up the list of features that are added! :/

**Comment by valstu** [comment]
I also feel this should have higher priority, I've had multiple projects where this exact feature would have been useful. I mean it is quite common case on any project to update multiple rows at once. @0x777 any idea if this will progress or is this abandoned?

**Comment by noel-dolan** [comment]
@valstu - for reference, I did manage to get around the problem by prefixing each mutation with an alias as shown above, so if you have control of the code you can dynamically do it that way.

**Comment by valstu** [comment]
I have considered that option as well even it might be a bit antipattern. In our case we're also generating TypeScript types from the mutations so I'm not sure how well this would work if mutations are dynamic, need to test this out.

**Comment by zbeyens** [comment]
My use-case is a bulk update from a list of object with variable length, so the above solution does not work. Same issue with the bulk delete.

**Comment by lilred** [comment]
Since upsert can't work against partial unique indexes this is causing issues with our ETL. I'd hate to have to go around Hasura and talk directly to Postgres, which up until now we've never had to do.

**Comment by shantiscript** [comment]
Hi am also i need of this, will likely go with multiple mutations within one call for now to get around limitations. But is a feature
that seems to be a very basic building block in the mutations toolset and agree with others on the priority of this

**Comment by mikolajkniejski** [comment]
Please add this feature!

**Comment by 0x777** [comment]
Hey folks, we will be picking this up soon. Can you share your use cases here? It'll be really helpful in designing the API.

**Comment by dariocravero** [comment]
@0x777 here's our use case. Say a list of fields need to be updated by PK, right now we do this:
```
mutation update($id1: uuid!, $object1: ..., $id2: uuid!, $object2: ...) {
  u1: update_users_by_pk(id: $id1, _set: $object2) { id }
  u2: update_users_by_pk(id: $id2, _set: $object2) { id }
}
```

Ideally, we'll have something like:
```
mutation update($objects: [{ id, _set }, {id, _set}...]) {
  update_users_many($objects) { returning { id } }
}
```

So effectively like inserting many but with a way to tell what you update.

Thanks for taking this on. It'll help DX a lot and make mutations safer! (Right now we construct that mutation on the fly)

**Comment by lxblvs** [comment]
We are also updating a big list of objects. I would prefer something like this:

```
mutation updateMultiple($itemUpdates: [item_update_input!]!) {
  update_items(updates: $itemUpdates) {
    affected_rows
  }
}
```

where the parameter is

```
itemUpdates = [
                   {where: {id: {_eq: 1}}, set: {name: "potato"}},
                   {where: {id: {_eq: 2}}, set: {name: "another potato"}},
                   {where: {name: {_eq: "not a potato"}}, set: {name: "yet another potato"}},
]
```

**Comment by adimyth** [comment]
> 

Exactly! None of the solutions above take into account the possibility of having variable number of updates required

**Comment by tjad** [comment]
@0x777 I like the idea of having an `[update_input!]!` syntax which contains `where` and `_set` ,  similar to your first suggestion
or https://github.com/hasura/graphql-engine/issues/2768#issuecomment-1081624310

This at least allows simple syntax for multiple updates within single transaction. We can easily create the list of update objects dynamically.

In terms of `rows affected`, I think it could return a value for each object/clause in order of the provided update objects

K.I.S.S

**Comment by eviefp** [comment]
We've gone through a few internal design iterations on this, and I'm here to report what we've landed on and ask for feedback.

## What I'm working on
I am currently working on implementing the version initially suggested by @0x7777, which is essentially a multi-record update by primary key.

Internally, we will make sure the keys don't repeat. If they do, we will use the last value (since it's a list of updates, we'll just pick the one that is closest to the list's end).

This will get translated into a single Postgres UPDATE statement. This is important, because this means we get the best possible performance.

## What were the designs we considered
One key technical fact is that Postgres will NOT update a row twice in the same statement. For example, say we wanted to allow this query:

```graphql
{
  update_user_many(
    updates: [
      { where: {id: {_gt: 1}, _set: {name: "hello"}},
      { where: {id: {_eq: 2}, _set: {description: "world"}}
    ]
  ) {
    affected_rows
  }
}
```

If we have a record with `id = 2`, then we can't know, before running the query, whether the first or the second update will trigger. However, only one of them will. This sort of non-determinism is not something we want to carry over to our product because it's unexpected and can easily introduce bugs in our user's code.

In the case of updates by primary key, it's fairly easy to detect when there's an overlap. However, as soon as we add more operators and other columns in the mix, the problem becomes incredibly complex (often times not solvable).

This means we end up with two options:
- the one we ended up picking: only allow equality on ID instead of a generalized `where` clause; this means very good performance and predictable results
- allow generic `where` clauses and run multiple updates in sequence; this means worse performance but allows overlapping/updating the same row multiple times, while keeping the latest value

### What about RETURNING?

In the primary key update version, returning is relatively simple to do and shouldn't surprise anyone.

However, the general `where` clause brings further complications: say we run two updates, and the first updates 2 rows, and the second 3. Should `affected_rows` be 5? Well, that might not be true because some of those rows could overlap. We could say "at least 2, at most 5". Or we could just return a list of `affected_rows` (one for each update).

Returning columns for affected rows for each operation can also be a bit tricky and impair performance.

## Conclusion

So, in conclusion, we're going for the solution that:
- is fast
- will be out ASAP for everyone to have a look at and give us feedback
- has predictable results

At the same time, we're wondering: how important is having a generic `where` clause? Are there any specific operators other than `eq` which you feel are essential for this feature?

**Comment by jackherizsmith** [comment]
Hello! I have just come across this thread, looks like great timing. Given the simplicity of option 1, the complexity of option 2 (i.e. conditionally selecting rows with `where`) and significance it would have for a lot of developers, I think this is the right solution for the product. For the scale of improvement we are about to experience, developers looking for an all in one solution can in the meantime simply run a prior query that returns the IDs they are looking to update.

**Comment by lxblvs** [comment]
This is very cool, thank you for this feature. I will definitely use it and praise the developers who wrote this code.

But I still hope that at some point you change your position on complex where clauses in the multiupdates. If we want to shoot ourselves in the foot and run conflicting mutations in a single call - please let us. Or maybe have a flag to run them sequentially and not as a single transaction? 

For now we can do a preflight call to resolve the ids or maybe have an Apollo preprocessor in front of Hasura that does it for us.

**Comment by revskill10** [comment]
Hi @eviefp It's almost the same as Prisma Transaction feature, but one more difference.

In Prisma transaction, they can mix match both query and mutation though.

**Comment by eviefp** [comment]
> But I still hope that at some point you change your position on complex where clauses in the multiupdates. If we want to shoot ourselves in the foot and run conflicting mutations in a single call - please let us.

We actually did change that! So right now we allow arbitrary `where` clauses that can freely overlap. We just run them in sequence, inside a transaction.

>  Or maybe have a flag to run them sequentially and not as a single transaction?

If there's user interest, we could definitely add a flag/option to allow running outside a transaction scope, ignoring errors.

**Comment by lxblvs** [comment]
@eviefp oh great! Are you also having your original `update_table_many_by_pk`? Running optimised mutations by pk would be quite badass for bulk deterministic mutations (like updating rows in a table where ids are known by the front-end)

**Comment by eviefp** [comment]
> @eviefp oh great! Are you also having your original `update_table_many_by_pk`? Running optimised mutations by pk would be quite badass for bulk deterministic mutations (like updating rows in a table where ids are known by the front-end)

@lxblvs We gave that up in favor of this current iteration. However, if we get enough requests, we can definitely prioritize the original `update_table_many_by_pk`!

**Comment by ajohnson1200** [comment]
> bulk deterministic mutations (like updating rows in a table where ids are known by the front-end)

Just so that we deeply understand the problem, can you talk through your use case in a little more detail?  ie: when you say "bulk", are you talking 10 rows, 100 rows, or 1000 rows?  And does the new solution that @eviefp outlined about prohibit you from accomplishing that goal... or is it instead that you can accomplish the goal, it's just not as fast?

**Comment by sassela** [comment]
Thanks for your patience with this, folks. I'm closing this issue as it was implemented at https://github.com/hasura/graphql-engine/commit/d76aab99e103adc7236b15f629839b34869ee944 and released in [v2.10.0](https://github.com/hasura/graphql-engine/releases/tag/v2.10.0). But please keep your feedback coming via comments on this issue; the Hasura team will be notified of it.
